src/basic/time-util.c

```diff
@@ -1329,6 +1329,94 @@ int get_timezones(char ***ret) {
         return 0;
 }
 
+static int get_timezones_from_tzdata_zi(char ***ret) {
+        _cleanup_fclose_ FILE *f = NULL;
+        _cleanup_strv_free_ char **zones = NULL;
+        size_t n_zones = 0;
+        int r;
+
+        assert(ret);
+
+        zones = strv_new("UTC");
+        if (!zones)
+                return -ENOMEM;
+
+        n_zones = 1;
+
+        f = fopen("/usr/share/zoneinfo/tzdata.zi", "re");
+        if (f) {
+                for (;;) {
+                        _cleanup_free_ char *line = NULL, *w = NULL;
+                        char *p, type;
+                        size_t k;
+
+                        r = read_line(f, LONG_LINE_MAX, &line);
+                        if (r < 0)
+                                return r;
+                        if (r == 0)
+                                break;
+
+                        p = strstrip(line);
+
+                        /* The only lines we care about are Zone and Link lines.
+                         * Zone lines have the timezone in the 2nd field.
+                         * Link lines have the timezone in the 3rd field.
+                         * See 'man zic' for more detail. */
+                        if (isempty(p) || !IN_SET(*p, 'Z', 'z', 'L', 'l'))
+                                continue;
+
+                        type = *p;
+
+                        /* move to second field */
+                        p += strcspn(p, WHITESPACE);
+                        p += strspn(p, WHITESPACE);
+
+                        if (type == 'L' || type == 'l') {
+                                /* move to third field */
+                                p += strcspn(p, WHITESPACE);
+                                p += strspn(p, WHITESPACE);
+                        }
+
+                        k = strcspn(p, WHITESPACE);
+                        if (k <= 0)
+                                continue;
+
+                        w = strndup(p, k);
+                        if (!w)
+                                return -ENOMEM;
+
+                        if (!GREEDY_REALLOC(zones, n_zones + 2))
```

Comment:
why +2?

Reply:
this was all copied from the existing get_timezones() that read the zone1970.tab file; i should review it for correctness

Reply:
it's +2 for the new timezone name, plus the trailing NULL, as n_zones doesn't count the null

Reply:
i redid it to use strv_extend instead, which seems better